src/cr/cube/matrix/cubemeasure.py

```diff
@@ -87,851 +94,1251 @@ def _slice_idx_expr(cls, cube, slice_idx):
+                valid_counts if valid_counts is not None else cube.unweighted_counts
+            )
+        diff_nans = True if valid_counts is not None else False
+        CubeCountsCls = (
+            _MrXMrCubeCounts
+            if dimension_type_strings == ("MR", "MR")
+            else _MrXArrCubeCounts
+            if dimension_type_strings == ("MR", "ARR")
+            else _MrXCatCubeCounts
+            if dimension_type_strings == ("MR", "CAT")
+            else _ArrXMrCubeCounts
+            if dimension_type_strings == ("ARR", "MR")
+            else _ArrXArrCubeCounts
+            if dimension_type_strings == ("ARR", "ARR")
+            else _ArrXCatCubeCounts
+            if dimension_type_strings == ("ARR", "CAT")
+            else _CatXMrCubeCounts
+            if dimension_type_strings == ("CAT", "MR")
+            else _CatXArrCubeCounts
+            if dimension_type_strings == ("CAT", "ARR")
+            else _CatXCatCubeCounts
+        )
+
+        return CubeCountsCls(
+            dimensions, counts[cls._slice_idx_expr(cube, slice_idx)], diff_nans
         )
 
     @lazyproperty
-    def means(self):
-        """2D np.float64 ndarray of cube means."""
+    def column_bases(self):
+        """2D np.float64 ndarray of column-wise bases for each valid matrix cell."""
         raise NotImplementedError(  # pragma: no cover
-            "`%s` must implement `.means`" % type(self).__name__
+            "`%s` must implement `.column_bases`" % type(self).__name__
         )
 
+    @lazyproperty
+    def columns_base(self):
+        """Optional 1D np.float64 ndarray of column-wise base for each valid column.
 
-class _CatXCatCubeMeans(_BaseCubeMeans):
-    """Means cube-measure for a slice with no MR dimensions."""
+        None if rows dimension is an ARRAY or MR because the base can vary among columns
+        """
+        return None
 
     @lazyproperty
-    def means(self):
-        """2D np.float64 ndarray of means for each valid matrix cell."""
-        return self._means
+    def columns_table_base(self):
+        """Optional 1D np.float64 ndarray of table-wise base for each valid column.
 
+        The name is a mouthful, but each component is meaningful.
+        - "columns": Indicates it is a marginal in the "columns" orientation (kind of
+        like a stripe in the shape of a row).
+        - "table": Indicates that it is the base for the whole table. When the
+        `.table_base` exists (CAT X CAT), it is a repetition of that, but when
+        the rows are array (and therefore we can't sum across them), each cell has
+        its own value.
+        - "base": Indicates that it is the base, not necessarily the counts (eg the sum
+        of selected and non-selected for MR variables)
 
-class _CatXMrCubeMeans(_BaseCubeMeans):
-    """Means cube-measure for a NOT_MR_X_MR slice.
+        None if rows dimension is an ARRAY or MR because the base can vary among columns.
+        """
+        return None
```

Comment:
This isn't making sense to me as stated. The whole point of this marginal is so the table-base can vary between (among) columns, right? Is what you mean to say: "`None` if rows-dimension is an ARRAY type because there is no single per-column table-base. In that case, each cell of the column can have a distinct table-base."

I suppose a quick re-word would be `s/among/within each`, but seems maybe worth it to spell things out a little more given this isn't so intuitive. Maybe also worth mentioning what this marginal is good for, like it forms the denominator in measure X when dimensions are Y or whatever it is, just to give some more possible grasping points.

Reply:
Yeah, there was an unfortunate typo in there, updated